### DISCUSSION TITLE
Add design related identifier to buttons and other navigation items

### DISCUSSION THREAD

**Discussion by SatishMulchandani** [apache/superset#25012] [Ideas]
In the current Superset version, we can change colors and fonts of buttons and navigation items by adding colors and typography in THEME_OVERRIDES object in superset_config.py. The problem with this approach is that we cannot overwrite other styles such as outline, border types to these buttons and navigation items. And also identifiers to distinguish different elements are missing for e.g. HTML tag and attribute of Create Dashboard and Bulk Select are same. So to apply different styles to these elements, I have to add them in their CSS in JS styled components. 

To implement custom styles, we are planning to update superset frontend code
1) Add identifiers to buttons and navigation items (for e.g. add class "btn-primary" to primary button)
2) Add our custom styles in theme.less file and import it in theme.ts file

theme.less file
.root {
    --primary-color: white;
    --primary-text-color: black;
}

button{
   &.btn-primary{
        background-color: @primary-color;
        color: @primary-text-color;
        border: 2px dashed @primary-border-color;   
   }
  &.disabled{
       background-color: @disabled-color;
       color: @disabled-text-color;
  }
}

Can someone confirm if I am on the right track or suggest any other alternative solution.

**Comment by natarajamanoj** [comment]
@geido Is this something you can help us with your recommendations? Or direct us with who can?

**Reply by geido** [reply]
Are you planning to contribute back those added classes to the open source project? I think that might be hard to justify as in most cases it will make the code harder to read and those classes will be pretty much arbitrary.

There are plans to support full themability in the future but I don't think it will happen soon.

You can probably stretch the CSS to customize parts of the application without additional classes.

**Reply by natarajamanoj** [reply]
I understand. We intend to contribute back to the community. Though our short term plans are to stretch CSS to achieve this need, we wanted to get along with community in building and contributing back.

Is there somewhere we can read about full themability?

**Reply by SatishMulchandani** [reply]
@geido, superset is using emotions to style individual parts. So to add custom styles (for e.g. outline) to certain buttons, we have to manually add those styles in the components where they are used. This manual task of adding style is very time consuming. Also, down the line if there is upgrade to superset, then all our individual styles will be lost. That's why we have proposed to add identifier to them so that we can apply styles at global level.

You have mentioned in your comment that full themability is planned in future release. We would like to know if there is any document or reference materials that is created for this. We would like to change our approach to superset future plan so that we have minimal changes in the future.

**Reply by geido** [reply]
@rusackas what are your thoughts on this topic?

**Reply by natarajamanoj** [reply]
@rusackas : Is this something you can share us with your thoughts?
